starboard/nplb/maximum_player_configuration_explorer.cc

```diff
@@ -0,0 +1,224 @@
+// Copyright 2023 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "starboard/nplb/maximum_player_configuration_explorer.h"
+
+#include <algorithm>
+#include <stack>
+#include <unordered_map>
+
+#include "starboard/common/log.h"
+#include "starboard/nplb/player_test_util.h"
+#include "starboard/player.h"
+
+namespace starboard {
+namespace nplb {
+
+namespace {
+
+class HashFunction {
+ public:
+  size_t operator()(const std::vector<int>& vector) const {
+    size_t hash = 0;
+    for (const auto& v : vector) {
+      hash += v;
+      hash *= 10;  // since we support at most 7 instances, 10 is good enough.
+    }
+    return hash;
+  }
+};
+
+}  // namespace
+
+bool PosetGreaterThanOrEqualTo(const std::vector<int>& a,
+                               const std::vector<int>& b) {
+  SB_DCHECK(a.size() == b.size());
+
+  for (int i = 0; i < a.size(); ++i) {
+    if (a[i] < b[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool PosetContainedByMaxElementsSet(
+    const std::set<std::vector<int>>& max_elements_set,
+    const std::vector<int>& v) {
+  for (const auto& max_element : max_elements_set) {
+    if (PosetGreaterThanOrEqualTo(max_element, v)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+std::set<std::vector<int>> ScheduleDFSOnGradedPoset(
+    int resource_types,
+    int max_instances_per_resource,
+    const std::function<bool(const std::vector<int>&)>& test_functor) {
+  SB_DCHECK(resource_types > 0);
+  SB_DCHECK(max_instances_per_resource > 0);
+  SB_DCHECK(resource_types <= 7 && max_instances_per_resource <= 7)
+      << "Exploring this solution space may be a time-consuming process.";
+  SB_DCHECK(test_functor);
+
+  std::stack<std::pair<std::vector<int>, int>> stack;
+  std::unordered_map<std::vector<int>, bool, HashFunction> advanceable;
+  std::set<std::vector<int>> max_elements_set;
+
+  stack.push(std::make_pair(std::vector<int>(resource_types, 0), 0));
+  advanceable[stack.top().first] = false;
+  while (!stack.empty()) {
```

Comment:
Yes, the search described is a depth-first search. The sibling nodes are encoded by index. For example, assume that there are two resources and each resource can support up to three instances. Assume the answer is the set {(2,0),(1,2),(0,3)}.

When we arrive at the node (2,0), the stack has three elements and looks like this:
``` 
((0, 0), index = 0), ((1, 0) , index = 0), ((2, 0) , index = 0)
```
We then check the children of (2,0), which are (3,0) and (2,1). Both of these nodes are not supported, so we backtrack to the previous level. The stack now looks like this:
```
((0, 0), index = 0), ((1, 0) , index = 1)
```
We then check the node (1,1), which is supported. The stack now looks like this:
```
((0, 0), index = 0), ((1, 0) , index = 1), ((1, 1) , index = 1)
```
We then move to the node (1,2), which is also supported. The stack now looks like this:
```
((0, 0), index = 0), ((1, 0) , index = 1), ((1, 1) , index = 1), ((1, 2) , index = 1)
```
We then check the node (1,3), which is a dead node. We then backtrack to the node (1,2), which has two children, (2,2) and (1,3). We know that (1,3) is a dead node, and we cannot find (2,2) in the set "advancable". 
The set "advancable" stores nodes that we have visited lexicographically. If a node is not in this set, it implies that there is a dead node in the path from the root node to it. Since the child node of a dead node must also be a dead node, we can deduce that (2,2) is a dead node.

Since (1,2) is the final live node and all of its children are dead nodes, it is a maximum element.

As described, we actually traverse the tree in depth-first order. However, by using the property of posets, we can avoid some unnecessary search steps. This is because creating and destroying players is expensive, and by using posets, we can avoid creating and destroying players that we know will not be supported.

Reply:
The main reason I am asking is, the "graph" here seems more like a tree to me, i.e. there is a clear root node for all sub-trees, and there isn't more than one connection from an upper level node to a lower level node.  The algorithm also looks like a pre-order traversal.

Reply:
Whatever the algorithm/data structure being stored here, it would be much more readable if it were abstracted away into a reusable class instead. That way, this class can use in a high level way and not get bogged down in the details.

Reply:
We should definitely make it the code more readable, but I also would avoid making it general reusable code for now, for two reasons:
1. It's for testing only, making it general purpose requires much more efforts.
2. The algorithm of this class has special requirements, i.e. it keeps searching down on the first edge of a node, instead of expanding into all connected nodes as a typical DFS does.  This is to minimize the creation/destroying of underlying resources.  A general implementation doesn't have to involve this kind of quirkiness.

Reply:
It doesn't necessarily need to be generally reusable, I'd just like it to be a separate class with a high level interface that looks something like "add node, delete node, search" so that calling functions are easy to read